src/dotnet/Users.Service/Controllers/MobileAuthController.cs

```diff
@@ -153,6 +157,66 @@ public async Task SignIn(string sessionId, string scheme, CancellationToken canc
         }
     }
 
+    [HttpPost("signInAppleWithCode")]
+    [Consumes("application/x-www-form-urlencoded")]
+    public async Task<IActionResult> SignInAppleWithCode(
+        [FromForm] IFormCollection request,
+        CancellationToken cancellationToken)
+    {
+        var code = request["Code"].ToString();
+        var sessionId = request["SessionId"].ToString();
+        var email = request["Email"].ToString();
+        var name = request["Name"].ToString();
+        var authenticationHandlerProvider = Services.GetRequiredService<IAuthenticationHandlerProvider>();
+        var userSettings = Services.GetRequiredService<UsersSettings>();
+        var handler = await authenticationHandlerProvider
+            .GetHandlerAsync(HttpContext, AppleAuthenticationDefaults.AuthenticationScheme)
+            .ConfigureAwait(false);
+        if (handler is not AppleAuthenticationHandler appleAuthHandler)
+            throw new InvalidOperationException($"{nameof(AppleAuthenticationHandler)} is not found.");
+
+        var options = appleAuthHandler.Options;
+        var context = new AppleGenerateClientSecretContext(HttpContext, appleAuthHandler.Scheme, options);
+        options.ClientId = userSettings.AppleAppId;
+        options.ClientSecret = await options.ClientSecretGenerator.GenerateAsync(context).ConfigureAwait(false);
+
+        using var tokenResponse = await ExchangeCodeAsync(code, options, cancellationToken).ConfigureAwait(false);
+        if (tokenResponse.Error != null) {
+            Logger.LogError(tokenResponse.Error, $"{nameof(SignInAppleWithCode)} error.");
+
+            return BadRequest();
+        }
+
+        var identity = new ClaimsIdentity(options.ClaimsIssuer);
+
+        if (!email.IsNullOrEmpty()) {
```

Comment:
I see that you take email and name from the client. 
For Google I take them from `tokenResponse`. Is it not possible to do it here the same way?

Reply:
It's the only place where we can get this data
Also it's available only at the very first request